queue_job/migrations/12.0.2.3.0/post-migration.py

```diff
@@ -0,0 +1,47 @@
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html)
+
+import logging
+
+from odoo import SUPERUSER_ID, api
+
+_logger = logging.getLogger(__name__)
+
+
+def migrate(cr, version):
+    with api.Environment.manage():
+        env = api.Environment(cr, SUPERUSER_ID, {})
+        _logger.info("Computing exception name for failed jobs")
+        _compute_jobs_new_values(env)
+
+
+def _compute_jobs_new_values(env):
+    for job in env["queue.job"].search(
+        [("state", "=", "failed"), ("exc_info", "!=", False)]
+    ):
+        exception_details = _get_exception_details(job)
+        if exception_details:
+            job.update(exception_details)
+
+
+def _get_exception_details(job):
+    for line in reversed(job.exc_info.splitlines()):
```

Comment:
I had to do this: 

```suggestion
    exc_info = job.read(fields=['exc_info'])[0]['exc_info']
    for line in reversed(exc_info.splitlines()):
```

Otherwise, browsing the job could fail if the job is on a model which is not yet loaded by the registry:

```
Traceback (most recent call last):
  File "/odoo/src/odoo/service/server.py", line 1162, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "/odoo/src/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/odoo/src/odoo/modules/loading.py", line 417, in load_modules
    force, status, report, loaded_modules, update_module, models_to_check)
  File "/odoo/src/odoo/modules/loading.py", line 313, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/odoo/src/odoo/modules/loading.py", line 227, in load_module_graph
    migrations.migrate_module(package, 'post')
  File "/odoo/src/odoo/modules/migration.py", line 190, in migrate_module
    migrate(self.cr, installed_version)
  File "/odoo/external-src/queue/queue_job/migrations/12.0.2.3.0/post-migration.py", line 14, in migrate
    _compute_jobs_new_values(env)
  File "/odoo/external-src/queue/queue_job/migrations/12.0.2.3.0/post-migration.py", line 21, in _compute_jobs_new_values
    exception_details = _get_exception_details(job)
  File "/odoo/external-src/queue/queue_job/migrations/12.0.2.3.0/post-migration.py", line 27, in _get_exception_details
    for line in reversed(job.exc_info.splitlines()):
  File "/odoo/src/odoo/fields.py", line 1036, in __get__
    self.determine_value(record)
  File "/odoo/src/odoo/fields.py", line 1139, in determine_value
    record._prefetch_field(self)
  File "/odoo/src/odoo/models.py", line 2880, in _prefetch_field
    result = records.read([f.name for f in fs], load='_classic_write')
  File "/odoo/src/odoo/models.py", line 2835, in read
    vals[name] = convert(record[name], record, use_name_get)
  File "/odoo/src/odoo/models.py", line 5167, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/odoo/src/odoo/fields.py", line 1044, in __get__
    return self.convert_to_record(value, record)
  File "/odoo/external-src/queue/queue_job/fields.py", line 64, in convert_to_record
    return json.loads(value or default, cls=JobDecoder, env=record.env)
  File "/usr/lib/python3.5/json/__init__.py", line 332, in loads
    return cls(**kw).decode(s)
  File "/usr/lib/python3.5/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.5/json/decoder.py", line 355, in raw_decode
    obj, end = self.scan_once(s, idx)
  File "/odoo/external-src/queue/queue_job/fields.py", line 107, in object_hook
    model = self.env[obj['model']]
  File "/odoo/src/odoo/api.py", line 831, in __getitem__
    return self.registry[model_name]._browse((), self)
  File "/odoo/src/odoo/modules/registry.py", line 177, in __getitem__
    return self.models[model_name]
KeyError: 'xxx.product.template'
```

Reply:
But we would better have to read all the "exc_info" at a time for all the jobs rather than 1 by 1 as I did here.

Reply:
Here is the better proposal reading all the rows at once: https://github.com/OCA/queue/pull/333/files#r626439498